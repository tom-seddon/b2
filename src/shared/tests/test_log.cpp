#include <shared/system.h>
#include <shared/log.h>
#include <shared/testing.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <thread>
#include <atomic>
#include <vector>

//////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////

static const char EXPECTED[] = {
#include "test_log.expected.txt"
};

static std::string g_str;
static LogPrinterString g_str_printer(&g_str);

LOG_DEFINE(TEST, "TEST", &g_str_printer);
LOG_DEFINE(OUT, "", &log_printer_stdout);

static int TestHighlightFn(size_t offset, void *data) {
    return offset == (uintptr_t)data;
}

static void TestDumpBytes(void) {
    LOG(TEST).SetPrefix("");

    const uint8_t data[] = {
        0x5B,
        0x5E,
        0x54,
        0x43,
        0x94,
        0xFB,
        0xB4,
        0x1D,
        0x58,
        0x7F,
        0xF9,
        0xD3,
        0x95,
        0xAB,
        0x26,
        0x64,
        0x66,
        0x80,
        0x2C,
        0x4A,
        0x95,
        0x9D,
        0x45,
        0x66,
        0xE7,
        0x89,
        0xB4,
        0x0B,
        0x09,
        0xBE,
        0x41,
        0x0D,
        0xD6,
        0xF0,
        0x0B,
        0xDF,
        0xBA,
        0x38,
        0xBE,
        0x2B,
        0xBE,
        0x94,
        0x5E,
        0xF8,
        0x8F,
        0x75,
        0x9E,
        0xF5,
        0x2C,
        0xE2,
        0xFD,
        0x4A,
        0x21,
        0x8F,
        0x86,
        0xF7,
        0x68,
        0x97,
        0xC5,
        0xEB,
        0x1B,
        0x7D,
        0x24,
        0x1A,
        0x55,
        0x58,
        0x17,
        0x0E,
        0xB3,
        0x3B,
        0x47,
        0x13,
        0x95,
        0x19,
        0xE5,
        0x1C,
        0xC7,
        0x7D,
        0x2E,
        0x8B,
        0x20,
        0xC4,
        0x86,
        0x83,
        0xCA,
        0xCD,
        0x29,
        0xE4,
        0x06,
        0xA7,
        0x4D,
        0xE3,
        0x1A,
    };

    const char data_expected[] =
        "0x00000000: 5B 5E 54 43 94 FB B4 1D 58 7F F9 D3 95 AB 26 64  [^TC....X.....&d\n"
        "0x00000010: 66 80 2C 4A 95 9D 45 66 E7 89 B4 0B 09 BE 41 0D  f.,J..Ef......A.\n"
        "0x00000020: D6 F0 0B DF BA 38 BE 2B BE 94 5E F8 8F 75 9E F5  .....8.+..^..u..\n"
        "0x00000030: 2C E2 FD 4A 21 8F 86 F7 68 97 C5 EB 1B 7D 24 1A  ,..J!...h....}$.\n"
        "0x00000040: 55 58 17 0E B3 3B 47 13 95 19 E5 1C C7 7D 2E 8B  UX...;G......}..\n"
        "0x00000050: 20 C4 86 83 CA CD 29 E4 06 A7 4D E3 1A            .....)...M..   \n";

    const char data_17_columns_expected[] =
        "0x00000000: 5B 5E 54 43 94 FB B4 1D 58 7F F9 D3 95 AB 26 64 66  [^TC....X.....&df\n"
        "0x00000011: 80 2C 4A 95 9D 45 66 E7 89 B4 0B 09 BE 41 0D D6 F0  .,J..Ef......A...\n"
        "0x00000022: 0B DF BA 38 BE 2B BE 94 5E F8 8F 75 9E F5 2C E2 FD  ...8.+..^..u..,..\n"
        "0x00000033: 4A 21 8F 86 F7 68 97 C5 EB 1B 7D 24 1A 55 58 17 0E  J!...h....}$.UX..\n"
        "0x00000044: B3 3B 47 13 95 19 E5 1C C7 7D 2E 8B 20 C4 86 83 CA  .;G......}.. ....\n"
        "0x00000055: CD 29 E4 06 A7 4D E3 1A                             .)...M..         \n";

    const char data_highlighted_expected[] =
        "0x00000000: 5B 5E 54 43 94 FB*B4 1D 58 7F F9 D3 95 AB 26 64  [^TC....X.....&d\n"
        "0x00000010: 66 80 2C 4A 95 9D 45 66 E7 89 B4 0B 09 BE 41 0D  f.,J..Ef......A.\n"
        "0x00000020: D6 F0 0B DF BA 38 BE 2B BE 94 5E F8 8F 75 9E F5  .....8.+..^..u..\n"
        "0x00000030: 2C E2 FD 4A 21 8F 86 F7 68 97 C5 EB 1B 7D 24 1A  ,..J!...h....}$.\n"
        "0x00000040: 55 58 17 0E B3 3B 47 13 95 19 E5 1C C7 7D 2E 8B  UX...;G......}..\n"
        "0x00000050: 20 C4 86 83 CA CD 29 E4 06 A7 4D E3 1A            .....)...M..   \n";

    const char data_first_address_expected[] =
        "0x1122334455667788: 5B 5E 54 43 94 FB B4 1D 58 7F F9 D3 95 AB 26 64  [^TC....X.....&d\n"
        "0x1122334455667798: 66 80 2C 4A 95 9D 45 66 E7 89 B4 0B 09 BE 41 0D  f.,J..Ef......A.\n"
        "0x11223344556677A8: D6 F0 0B DF BA 38 BE 2B BE 94 5E F8 8F 75 9E F5  .....8.+..^..u..\n"
        "0x11223344556677B8: 2C E2 FD 4A 21 8F 86 F7 68 97 C5 EB 1B 7D 24 1A  ,..J!...h....}$.\n"
        "0x11223344556677C8: 55 58 17 0E B3 3B 47 13 95 19 E5 1C C7 7D 2E 8B  UX...;G......}..\n"
        "0x11223344556677D8: 20 C4 86 83 CA CD 29 E4 06 A7 4D E3 1A            .....)...M..   \n";

    g_str.clear();
    LogDumpBytes(&LOG(TEST), data, sizeof data);
    TEST_EQ_SS(g_str, data_expected);

    g_str.clear();
    {
        LogDumpBytesExData d = {};

        d.num_dump_columns = 17;

        LogDumpBytesEx(&LOG(TEST), data, sizeof data, &d);
    }
    TEST_EQ_SS(g_str, data_17_columns_expected);

    g_str.clear();
    {
        LogDumpBytesExData d = {};

        d.highlight_fn = &TestHighlightFn;
        d.highlight_data = (void *)(uintptr_t)5;

        LogDumpBytesEx(&LOG(TEST), data, sizeof data, &d);
    }
    TEST_EQ_SS(g_str, data_highlighted_expected);

    g_str.clear();
    {
        LogDumpBytesExData d = {};

        d.first_address = 0x1122334455667788;

        LogDumpBytesEx(&LOG(TEST), data, sizeof data, &d);
    }
    TEST_EQ_SS(g_str, data_first_address_expected);
}

int main(void) {
    LOGF(TEST, "Line 1\nLine 2\n");
    LOGF(TEST, "Ordinary line\n\tTabbed line\n");

    LOGF(TEST, "Indent prefix: ");
    {
        LOGI(TEST);
        LOGF(TEST, "Indented line 1\n");
        LOGF(TEST, "Indented line 2\n");
        LOGF(TEST, "Indented line 3\n");

        LOGF(TEST, "Extra indent: ");

        {
            LOGI(TEST);
            LOGF(TEST, "Indented line 1\n");
            LOGF(TEST, "Indented line 2\n");
            LOGF(TEST, "Indented line 3\n");
        }
    }

    {
        std::string big_string;
        int num_lines = 0;

        while (big_string.size() < 9900) {
            big_string += "Line " + std::to_string(1 + num_lines) + ": ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n";
            ++num_lines;
        }

        big_string += "***\n";

        LOGF(TEST, "%s", big_string.c_str());

        if (big_string.size() < Log::PRINTF_BUFFER_SIZE) {
            LOGF(TEST, "NOTE: string was too small to exercise code for printing large strings.\n");
        }
    }

    TEST_EQ_SS(g_str, EXPECTED);

    LOG(TEST).EnsureBOL();
    g_str.clear();

    LOG(TEST).SetPrefix("");
    LogStringPrintable(&LOG(TEST), "Hello\n\r\t\bHello");
    LOG(TEST).Flush();

    /* LogDumpBytes(&LOG(OUT),g_buffer.text,g_buffer.size); */

    TEST_EQ_SS(g_str, "Hello\\n\\r\\t\\bHello");

    TestDumpBytes();

    return 0;
}
