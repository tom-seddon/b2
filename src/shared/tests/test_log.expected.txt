"TEST: Line 1\n"
"TEST: Line 2\n"
"TEST: Ordinary line\n"
"      Tabbed line\n"
"TEST: Indent prefix: Indented line 1\n"
"TEST:                Indented line 2\n"
"TEST:                Indented line 3\n"
"TEST:                Extra indent: Indented line 1\n"
"TEST:                              Indented line 2\n"
"TEST:                              Indented line 3\n"
"TEST: Line 1: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 2: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 3: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 4: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 5: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 6: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 7: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 8: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 9: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 10: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 11: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 12: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 13: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 14: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 15: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 16: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 17: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 18: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 19: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 20: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 21: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 22: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 23: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 24: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 25: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 26: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 27: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 28: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 29: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 30: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 31: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 32: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 33: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 34: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 35: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 36: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 37: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 38: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 39: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 40: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 41: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 42: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 43: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 44: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 45: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 46: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 47: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 48: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 49: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 50: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 51: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 52: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 53: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 54: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 55: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 56: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 57: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 58: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 59: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 60: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 61: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 62: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 63: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 64: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 65: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 66: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 67: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 68: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 69: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 70: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 71: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 72: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 73: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 74: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 75: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 76: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 77: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 78: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 79: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 80: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 81: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 82: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 83: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 84: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 85: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 86: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 87: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 88: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 89: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 90: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 91: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 92: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 93: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 94: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 95: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 96: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 97: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 98: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 99: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 100: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 101: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 102: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 103: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 104: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 105: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 106: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 107: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 108: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 109: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 110: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 111: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 112: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 113: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 114: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 115: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 116: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 117: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 118: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 119: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 120: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 121: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 122: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 123: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 124: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 125: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 126: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 127: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 128: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 129: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 130: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 131: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 132: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 133: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 134: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 135: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 136: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 137: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 138: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 139: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 140: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 141: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 142: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 143: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 144: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 145: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 146: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 147: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 148: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 149: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 150: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 151: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 152: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 153: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 154: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 155: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 156: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: Line 157: ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz\n"
"TEST: ***\n"
